akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -171,43 +205,67 @@ import scala.util.control.NonFatal
           else if (t == null) throw ex
         case s: Subscription ⇒
+          if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($s).onError(${t.getMessage}) -> ErrorPublisher")
           if (!compareAndSet(s, ErrorPublisher(ex, "failed-VirtualProcessor"))) rec(ex)
           else if (t == null) throw ex
         case Both(s) ⇒
+          if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode(Both($s)).onError(${t.getMessage}) -> ErrorPublisher")
           set(Inert)
           try tryOnError(s, ex)
           finally if (t == null) throw ex // must throw NPE, rule 2:13
         case s: Subscriber[_] ⇒ // spec violation
+          if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($s).onError(${t.getMessage}) -> Inert")
           getAndSet(Inert) match {
             case Inert ⇒ // nothing to be done
             case _     ⇒ ErrorPublisher(ex, "failed-VirtualProcessor").subscribe(s)
           }
-        case _ ⇒ // spec violation or cancellation race, but nothing we can do
+        case Establishing(s) ⇒
+          // keep trying until subscription established and can complete it
+          if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($s).onError(${t.getMessage}), loop")
+
+        case other ⇒
+          if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($other).onError(${t.getMessage}). spec violation or cancellation race")
+        // spec violation or cancellation race, but nothing we can do
       }
 
     val ex = if (t == null) exceptionMustNotBeNullException else t
     rec(ex)
   }
 
-  @tailrec override final def onComplete(): Unit =
+  @tailrec override final def onComplete(): Unit = {
     get() match {
-      case null            ⇒ if (!compareAndSet(null, EmptyPublisher)) onComplete()
-      case s: Subscription ⇒ if (!compareAndSet(s, EmptyPublisher)) onComplete()
-      case Both(s) ⇒
+      case null ⇒
+        if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode(null).onComplete -> EmptyPublisher")
+        if (!compareAndSet(null, EmptyPublisher)) onComplete()
+      case s: Subscription ⇒
+        if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($s).onComplete -> EmptyPublisher")
+        if (!compareAndSet(s, EmptyPublisher)) onComplete()
+      case b @ Both(s) ⇒
+        if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($s).onComplete -> Inert")
         set(Inert)
         tryOnComplete(s)
       case s: Subscriber[_] ⇒ // spec violation
+        if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($s).onComplete -> Inert")
         set(Inert)
         EmptyPublisher.subscribe(s)
-      case _ ⇒ // spec violation or cancellation race, but nothing we can do
+      case Establishing(s) ⇒
+        // keep trying until subscription established and can complete it
+        if (VirtualProcessor.Debug) println(s"VirtualPublisher#$hashCode($s).onComplete, loop")
+        onComplete()
```

Comment:
are we sure that will complete?
is there a risk of live lock (starvation) if several threads ends up spinning here and there is no thread remaining to complete the subscription?

Reply:
The only situation it wouldn't complete is if the establishing subscription never completes nor fails, I'm not sure what scenario that would be (or what we could possibly do about it).

I don't know the RS spec good enough to answer the part about several threads. A can several threads call try to complete the same subscriber? @ktoso ?

Reply:
I think it's dangerous to spin and depend on that another thread unlocks the spinning. See for example this discussion https://groups.google.com/d/msg/akka-user/dRxTX8WmWlA/ku-nM0KcHJ8J and https://github.com/akka/akka/issues/17216#issuecomment-93945089

Reply:
Thanks for digging that up, had no idea.

I think the solution would be to buffer onComplete, if establishing, rather than spinning, and then having establish trigger the complete if needed when done.

Reply:
It would be illegal to not complete establishing a sub, but it could happen; good to buffer onComplete instead